Generator/CompiledInterceptor.php

```diff
@@ -619,16 +619,54 @@ private function getCompiledMethodInfo(\ReflectionMethod $method, $config)
             $body[] = "}";
         }
 
+        $returnTypeValue = null;
         $returnType = $method->getReturnType();
-        $returnTypeValue = $returnType
-            ? ($returnType->allowsNull() ? '?' : '') . $returnType->getName()
-            : null;
-        if ($returnTypeValue === 'self') {
-            $returnTypeValue = $method->getDeclaringClass()->getName();
+        $isUnionReturnType = ($method->hasReturnType() && $method->getReturnType() instanceof \ReflectionUnionType);
+        if ($isUnionReturnType) {
+            foreach ($method->getReturnType()->getTypes() as $methodReturnTypesItem) {
+                $methodReturnTypeValue = $methodReturnTypesItem
+                    ? ($methodReturnTypesItem->getName() !== 'null' && $methodReturnTypesItem->allowsNull() ? '?' : '') . $methodReturnTypesItem->getName()
+                    : null;
+                if ($methodReturnTypeValue === 'self') {
+                    $methodReturnTypeValue = $methodReturnTypesItem->getDeclaringClass()->getName();
+                }
+                if (!$returnTypeValue) {
+                    $returnTypeValue = $methodReturnTypeValue;
+                } else {
+                    $returnTypeValue .= '|' . $methodReturnTypeValue;
+                }
+            }
+        }
+        
+        $isIntersectionType = ($method->hasReturnType() && $method->getReturnType() instanceof \ReflectionIntersectionType);
+        if ($isIntersectionType) {
+            foreach ($method->getReturnType()->getTypes() as $methodReturnTypesItem) {
+                $methodReturnTypeValue = $methodReturnTypesItem
+                    ? ($methodReturnTypesItem->getName() !== 'null' && $methodReturnTypesItem->allowsNull() ? '?' : '') . $methodReturnTypesItem->getName()
+                    : null;
+                if ($methodReturnTypeValue === 'self') {
+                    $methodReturnTypeValue = $methodReturnTypesItem->getDeclaringClass()->getName();
+                }
+                if (!$returnTypeValue) {
+                    $returnTypeValue = $methodReturnTypeValue;
+                } else {
+                    $returnTypeValue .= '&' . $methodReturnTypeValue;
+                }
+            }
         }
+        
+        if (!$returnTypeValue) {
+            $returnTypeValue = $returnType
+                ? ($returnType->allowsNull() ? '?' : '') . $returnType->getName()
+                : null;
+            if ($returnTypeValue === 'self') {
+                $returnTypeValue = $method->getDeclaringClass()->getName();
+            }
```

Comment:
This isn't strictly related to the issues this PR fixes, but @thlassche in #19 suggests adding this additional condition to properly handle `mixed` return type declarations, like `\Magento\Config\Console\Command\LocaleEmulator::emulate` in 2.4.7:
```
+            if ($returnType && $returnType->getName() === 'mixed') {
+                $returnTypeValue = 'mixed';
+            }
```